### DISCUSSION TITLE
Encapsulate nicegui.js

### DISCUSSION THREAD

**Discussion by falkoschindler** [zauberzeug/nicegui#3007] [Ideas / Feature Requests]
As noticed by @nigelsim in https://github.com/zauberzeug/nicegui/discussions/2984#discussioncomment-9302034, nicegui.js currently populates the global `window` scope, which can lead to conflicts with other JavaScript code like browser extensions. 

> Typically, I've see JS applications just encapsulate their own code inside a self executing block, so that this can't happen.

Even though I like the idea of wrapping nicegui.js in something like
```js
(function() {

})();
```
we need to expose some of the functions like `parseElements` or `createApp` which are used in index.html.

Maybe we can implement nicegui.js as a module, _exporting_ individual functions?

And maybe we can use bundlers like webpack or rollup to even further improve nicegui.js, e.g. by minifying the code or by bundling with Vue, Quasar and Tailwind?

Any help would be greatly appreciated, as this is not our primary area of expertise.

**Comment by nigelsim** [comment]
I've made a start [here](https://github.com/nigelsim/nicegui/tree/encapsulate-nicegui-js). A few manual tests and it looks fine, but I'm still figuring out the selenium tests before I put in a PR.

**Reply by nigelsim** [reply]
And [another approach](https://github.com/nigelsim/nicegui/tree/encapsulate-nicegui-js-modules) using JS modules, which I think is the better approach given NiceGUI is already using JS modules.
